velox/common/caching/SsdCache.cpp

```diff
@@ -0,0 +1,505 @@
+        auto entry = pins[i].entry();
+        entry->setSsdFile(this, offset);
+        char first = entry->tinyData() ? entry->tinyData()[0]
+                                       : entry->data().runAt(0).data<char>()[0];
+        auto size = entry->size();
+        SsdKey key = {
+            entry->key().fileNum, static_cast<uint64_t>(entry->offset())};
+        entries_[std::move(key)] = SsdRun(offset, size);
+        if (FLAGS_verify_ssd_write) {
+          verifyWrite(*entry, SsdRun(offset, size));
+        }
+        offset += size;
+        ++stats_.entriesWritten;
+        stats_.bytesWritten += size;
+      }
+    }
+    storeIndex += numWritten;
+  }
+}
+
+char* readBytes(int fd, int64_t offset, int32_t size) {
+  char* data = (char*)malloc(size);
+  pread(fd, data, size, offset);
+  return data;
+}
+
+namespace {
+int32_t indexOfFirstMismatch(char* x, char* y, int n) {
+  for (auto i = 0; i < n; ++i) {
+    if (x[i] != y[i]) {
+      return i;
+    }
+  }
+  return -1;
+}
+} // namespace
+
+void SsdFile::verifyWrite(AsyncDataCacheEntry& entry, SsdRun ssdRun) {
+  auto testData = std::make_unique<char[]>(entry.size());
+  auto rc = pread(fd_, testData.get(), entry.size(), ssdRun.offset());
+  VELOX_CHECK_EQ(rc, entry.size());
+  if (entry.tinyData()) {
+    if (0 != memcmp(testData.get(), entry.tinyData(), entry.size())) {
+      VELOX_FAIL("bad read back");
+    }
+  } else {
+    auto& data = entry.data();
+    int64_t bytesLeft = entry.size();
+    int64_t offset = 0;
+    for (auto i = 0; i < data.numRuns(); ++i) {
+      auto run = data.runAt(i);
+      auto compareSize = std::min<int64_t>(bytesLeft, run.numBytes());
+      auto badIndex = indexOfFirstMismatch(
+          run.data<char>(), testData.get() + offset, compareSize);
+      if (badIndex != -1) {
+        VELOX_FAIL("Bad read back");
+      }
+      bytesLeft -= run.numBytes();
+      offset += run.numBytes();
+      if (bytesLeft <= 0) {
+        break;
+      };
+    }
+  }
+}
+
+void SsdFile::updateStats(SsdCacheStats& stats) {
+  stats.entriesWritten += stats_.entriesWritten;
+  stats.bytesWritten += stats_.bytesWritten;
+  stats.entriesRead += stats_.entriesRead;
+  stats.bytesRead += stats_.bytesRead;
+  stats.entriesCached += entries_.size();
+  for (auto& regionSize : regionSize_) {
+    stats.bytesCached += regionSize;
+  }
+}
+
+void SsdFile::clear() {
+  std::lock_guard<std::mutex> l(mutex_);
+  entries_.clear();
+  std::fill(regionSize_.begin(), regionSize_.end(), 0);
+  writableRegions_.resize(numRegions_);
+  std::iota(writableRegions_.begin(), writableRegions_.end(), 0);
+  std::fill(regionScore_.begin(), regionScore_.end(), 0);
+}
+
+SsdCache::SsdCache(
+    std::string_view filePrefix,
+    uint64_t maxBytes,
+    int32_t numShards)
+    : filePrefix_(filePrefix),
+      numShards_(numShards),
+      groupStats_(std::make_unique<FileGroupStats>()) {
+  files_.reserve(numShards_);
+  uint64_t kSizeQuantum = numShards_ * SsdFile::kRegionSize;
```

Comment:
kSizeQuantum name suggests that it is a static constant, but it is not. Perhaps, rename.

Reply:
probably `uint64_t minQuantumSize`? defined as the size where the number of files having at least one region? Example: if num files = 10, then `minQuantumSize = 640 MB` since `kRegionSize = 1 << 26; // 64MB`